packages/shared/src/effect.ts

```diff
@@ -31,68 +31,67 @@ export const fetchEff = (
   input: RequestInfo | URL,
   init?: RequestInit,
 ): Effect.Effect<ResponseEsque, FetchError, FetchContext> =>
-  FetchContext.pipe(
-    Effect.andThen(({ fetch, baseHeaders }) =>
-      Effect.tryPromise({
-        try: () =>
-          fetch(input, {
-            ...init,
-            headers: {
-              ...filterObjectValues(baseHeaders, (v): v is string => v != null),
-              ...init?.headers,
-            },
-          }),
-        catch: (error) => new FetchError({ error, input }),
+  Effect.flatMap(FetchContext, ({ fetch, baseHeaders }) => {
+    const req = new Request(input, {
+      ...init,
+      headers: {
+        ...filterObjectValues(baseHeaders, (v): v is string => v != null),
+        ...init?.headers,
+      },
+    });
+    return Effect.tryPromise({
+      try: () => fetch(req.clone()),
+      catch: (error) => new FetchError({ error, input: req.clone() }),
+    }).pipe(
+      Effect.map((res) => Object.assign(res.clone(), { request: req.clone() })),
+      Effect.withSpan("fetch", {
+        attributes: { input: JSON.stringify(input) },
       }),
-    ),
-    Effect.withSpan("fetch", {
-      attributes: { input: JSON.stringify(input) },
-    }),
-  );
-
-// TODO: rename the other one to fetchEffJsonSchema and this to fetchEffJson
-// though generally I would avoid XandY kind utils in favor of composition
-export const fetchEffUnknown = (
-  input: RequestInfo | URL,
-  /** Schema to be used if the response returned a 2xx  */
-  init?: RequestInit,
-): Effect.Effect<
-  { json: unknown; ok: boolean; status: number },
-  FetchError | BadRequestError,
-  FetchContext
-> => {
-  const requestUrl =
-    typeof input === "string"
-      ? input
-      : input instanceof URL
-        ? input.toString()
-        : input.url;
+    );
+  });
 
-  return fetchEff(input, init).pipe(
-    Effect.andThen((res) =>
-      Effect.tryPromise({
-        try: async () => {
-          const json = await res.json();
-          return { json, ok: res.ok, status: res.status };
-        },
-        catch: (error) => new FetchError({ error, input }),
-      }),
-    ),
+export const json: (res: ResponseEsque) => Effect.Effect<
+  {
+    json: unknown;
+    ok: boolean;
+    status: number;
+  },
+  FetchError | BadRequestError<Json>,
+  never
+> = (res: ResponseEsque) =>
+  Effect.tryPromise({
+    try: async () => {
+      const json = await res.json();
+      return { json, ok: res.ok, status: res.status };
+    },
+    catch: (error) => new FetchError({ error, input: res.request }),
+  }).pipe(
     Effect.filterOrFail(
       ({ ok }) => ok,
       ({ json, status }) =>
         new BadRequestError({
-          input,
+          input: res.request,
```

Comment:
Have to double check what value we're actually getting from attaching this property, i dont think we use anything other than the url.

We already track it on the span in the base fetchEff function, so for debuggability it feels like we should improve our usage of those before starting drilling these props onto a bunch of errors🤔

Reply:
It depends what kind of debug story you're trying to achieve. I agree that the errors could be better described, I don't think that just recording the input brings us much value..

What do you think I should do? remove it altogether?